packages/@lwc/shared/src/attr-reflection.ts

```diff
@@ -0,0 +1,137 @@
+        if (normalizedValue !== currentValue) {
+            this.setAttribute(attrName, normalizedValue);
+        }
+    },
+});
+
+/** Descriptor for a boolean that checks for `attr="true"` or `attr="false"`, e.g. `spellcheck` and `draggable`. */
+const explicitBooleanDescriptor = (
+    attrName: string,
+    defaultValue: boolean
+): TypedPropertyDescriptor<boolean> => ({
+    configurable: true,
+    enumerable: true,
+    get(this: LightningElement): boolean {
+        const value = this.getAttribute(attrName);
+        return value === null ? defaultValue : value === String(defaultValue);
+    },
+    set(this: LightningElement, newValue: boolean): void {
+        const currentValue = this.getAttribute(attrName);
+        const normalizedValue = String(Boolean(newValue));
+        if (normalizedValue !== currentValue) {
+            this.setAttribute(attrName, normalizedValue);
+        }
+    },
+});
+
+/**
+ * Descriptor for a "true" boolean attribute that checks solely for presence, e.g. `hidden`.
+ */
+const booleanAttributeDescriptor = (attrName: string): TypedPropertyDescriptor<boolean> => ({
+    configurable: true,
+    enumerable: true,
+    get(this: LightningElement): boolean {
+        return this.hasAttribute(attrName);
+    },
+    set(this: LightningElement, newValue: boolean): void {
+        const hasAttribute = this.hasAttribute(attrName);
+        if (newValue) {
+            if (!hasAttribute) {
+                this.setAttribute(attrName, '');
+            }
+        } else {
+            if (hasAttribute) {
+                this.removeAttribute(attrName);
+            }
+        }
+    },
+});
+
+/**
+ * Descriptor for ARIA reflections, e.g. `ariaLabel` and `role`.
+ */
+const ariaDescriptor = (attrName: string): TypedPropertyDescriptor<string | null> => ({
+    configurable: true,
+    enumerable: true,
+    get(this: LightningElement): string | null {
+        return this.getAttribute(attrName);
+    },
+    set(this: LightningElement, newValue: string | null): void {
+        const currentValue = this.getAttribute(attrName);
+        if (newValue !== currentValue) {
+            // TODO [#3284]: According to the spec, IDL nullable type values
+            // (null and undefined) should remove the attribute; however, we
+            // only do so in the case of null for historical reasons.
+            if (isNull(newValue)) {
+                this.removeAttribute(attrName);
+            } else {
+                this.setAttribute(attrName, toString(newValue));
+            }
+        }
+    },
+});
+
+const descriptors: Record<string, PropertyDescriptor> = {
+    accessKey: stringDescriptor('accesskey'),
+    dir: stringDescriptor('dir'),
+    draggable: explicitBooleanDescriptor('draggable', true),
+    hidden: booleanAttributeDescriptor('hidden'),
+    id: stringDescriptor('id'),
+    lang: stringDescriptor('lang'),
+    spellcheck: explicitBooleanDescriptor('spellcheck', false),
```

Comment:
`spellcheck` needs some special handling; this is why `global-html-attributes/index.js` is still failing. Can be done in a follow-up PR.